File: matlab/demos/demo_genChainSpcs.m
%% demo_genChainSpcs.m
%  Spectral pitch class similarity (SPCS) of generator-chain tunings
%  against a reference chord, as the generator interval is swept
%  continuously.
%
%  A generator-chain of cardinality n is a scale formed by stacking n
%  copies of a generating interval, reduced modulo the period:
%    pitches = mod((0:n-1) * gen, period)
%
%  For each generator size (swept in fine steps from 0 to period/2), the
%  SPCS between the resulting scale and a reference chord is computed.
%  Results are reflected to fill the full range [0, period), since a
%  generator of size gen produces the same scale as period - gen.
%  The plot reveals which generator sizes produce scales that are most
%  similar to the reference chord — peaks correspond to tunings whose
%  intervals closely approximate those of the chord.
%
%  Examples of this type of plot appear as Examples 6.4 and 6.5 / Figures
%  5-7 in:
%    Milne, A. J., Sethares, W. A., Laney, R., & Sharp, D. B. (2011).
%    Modelling the similarity of pitch collections with expectation tensors.
%    Journal of Mathematics and Music, 5(1), 1-20.
%  (That paper uses the term "beta-chain" for what is here called a
%  generator-chain, with "beta" denoting the generator interval.)
%
%  Uses: batchCosSimExpTens, cosSimExpTens
%  (from the Music Perception Toolbox).

%% === User-adjustable parameters ===

% Reference chord (in cents)
%   4:5:6 JI major triad: [0, 386.31, 701.96] with period 1200
%   Bohlen-Pierce "major" triad: [0, 884.36, 1466.87] with period 1902
refPitches = [0, 1200*log2(5/4), 1200*log2(6/4)];
refWeights = [];   % weights for reference pitches (empty = uniform)
refName    = '4:5:6 JI major triad';

% Generator-chain parameters
nTones  = 19;      % number of tones in the chain
genStep = 0.1;     % step size for generator sweep (cents)

% Expectation tensor parameters
sigma  = 10;      % Gaussian smoothing width (cents)
r      = 2;       % dyad expectation tensor
isRel  = 1;       % relative (transposition-invariant)
isPer  = 1;       % periodic (pitch-class equivalence)
period = 1200;    % period in cents (1200 = octave; 1902 = tritave)

% Display options
invertCircular = true;   % if true, higher similarity is towards the centre
                         % (as in the article, which plots distance);
                         % if false, higher similarity is at the perimeter
nLabels        = 15;     % maximum number of peak labels on each plot
minLabelSpacing = 9;    % minimum spacing between labels (cents); increase
                         % to reduce clutter, decrease to show more peaks

%% === Build pitch matrices (0 to period/2 only) ===
% A generator-chain with generator gen produces the same scale as one with
% generator period - gen (stacking upward vs downward), so we compute only
% the first half and reflect the results.

halfRange = 0:genStep:(period / 2);
nHalf     = numel(halfRange);

% Reference chord: same for every row
pitchesA = repmat(refPitches, nHalf, 1);

% Reference weights
if ~isempty(refWeights)
    weightsA = repmat(refWeights, nHalf, 1);
else
    weightsA = [];
end

% Generator-chain pitch sets: each row is a different generator value
pitchesB = NaN(nHalf, nTones);
for i = 1:nHalf
    gen = halfRange(i);
    pitchesB(i, :) = mod((0:nTones-1) * gen, period);
end

%% === Compute similarities ===

fprintf('Computing SPCS of %d-tone generator-chains (gen = 0 to %.1f, step %.2f) against %s...\n', ...
    nTones, period / 2, genStep, refName);
opts = {'verbose', false};
if ~isempty(weightsA)
    opts = [{'weightsA', weightsA}, opts];
end
sHalf = batchCosSimExpTens(pitchesA, pitchesB, ...
    sigma, r, isRel, isPer, period, opts{:});
fprintf('Done.\n');

% Round to 3 decimal places for display (avoids floating-point artifacts
% such as values marginally above 1 or below 0)
sHalf = round(sHalf, 3);

% Reflect to produce the full range [0, period)
% gen and (period - gen) give identical scales, so the second half is
% simply sHalf reversed, excluding both endpoints to avoid duplication.
genRange = 0:genStep:(period - genStep);
s = [sHalf(:); flipud(sHalf(2:end-1))];

% If genStep does not divide period/2 evenly, the reflected vector may be
% slightly shorter or longer than genRange. Pad or trim to match.
if numel(s) ~= numel(genRange)
    warning(['genStep (%.4g) does not divide period/2 (%.4g) evenly — ' ...
        'steps will be slightly uneven near the midpoint.'], genStep, period / 2);
    if numel(s) < numel(genRange)
        s(end+1:numel(genRange)) = s(end);
    else
        s = s(1:numel(genRange));
    end
end

%% === Find peaks ===

[pks, locs] = findpeaks(s, 'MinPeakProminence', 0.01, 'SortStr', 'descend');

%% === Linear plot ===

figure('Name', 'Generator-chain SPCS', ...
       'Position', [100, 100, 1000, 400]);

plot(genRange, s, 'Color', [0.2 0.2 0.6], 'LineWidth', 0.8);
xlabel('Generator (cents)');
ylabel('Spectral pitch class similarity');
title(sprintf(['SPCS of %d-tone generator-chains with %s\n' ...
    '(r = %d, isRel = %d, \\sigma = %d cents, period = %d cents)'], ...
    nTones, refName, r, isRel, sigma, period));
xlim([0, period]);

hold on;
labelledGens = [];
labelled = 0;
for li = 1:numel(pks)
    genVal = genRange(locs(li));

    % Skip if too close to an already-labelled peak
    if ~isempty(labelledGens) && any(abs(labelledGens - genVal) < minLabelSpacing)
        continue;
    end

    plot(genVal, pks(li), 'r.', 'MarkerSize', 3);
    text(genVal, pks(li), sprintf('  %.1f', genVal), ...
        'FontSize', 7, 'Color', [0.8 0 0], ...
        'HorizontalAlignment', 'left', ...
        'VerticalAlignment', 'bottom');

    labelledGens = [labelledGens, genVal]; %#ok<AGROW>
    labelled = labelled + 1;
    if labelled >= nLabels
        break;
    end
end
hold off;

%% === Circular plot ===

% Map generator to angle: 0 at the top (12 o'clock), increasing clockwise
theta = 2 * pi * genRange / period;

% Close the curve by appending the first point
thetaC = [theta(:); theta(1)];
sC     = [s(:); s(1)];

figure('Name', 'Generator-chain SPCS (circular)', ...
       'Position', [100, 550, 600, 600]);

% SPCS as radius, scaled to fixed [0, 1] range:
%   invertCircular: SPCS = 0 at circumference (r = 1), SPCS = 1 at centre (r = 0)
%   normal:         SPCS = 0 at centre (r = 0), SPCS = 1 at circumference (r = 1)
if invertCircular
    rC = 1 - sC;
else
    rC = sC;
end

polarplot(thetaC, rC, 'Color', [0.2 0.2 0.6], 'LineWidth', 1);
ax = gca;
ax.ThetaZeroLocation = 'top';
ax.ThetaDir = 'clockwise';
ax.RLim = [0, 1];

% Tick labels in cents rather than degrees
nTicks = 12;
tickAngles = linspace(0, 360 - 360/nTicks, nTicks);
tickLabels = arrayfun(@(a) sprintf('%.0f', a * period / 360), ...
    tickAngles, 'UniformOutput', false);
ax.ThetaTick = tickAngles;
ax.ThetaTickLabel = tickLabels;

% Hide radial tick labels (the absolute radius values are not meaningful)
ax.RTickLabel = [];

if invertCircular
    titleExtra = 'centre = most similar';
else
    titleExtra = 'perimeter = most similar';
end
title(sprintf(['SPCS of %d-tone generator-chains with %s\n' ...
    '(r = %d, isRel = %d, \\sigma = %d cents, period = %d cents; %s)'], ...
    nTones, refName, r, isRel, sigma, period, titleExtra));

% Label top peaks on the circular plot
hold on;
labelledGensC = [];
labelledC = 0;
for li = 1:numel(pks)
    genVal = genRange(locs(li));
    if ~isempty(labelledGensC) && any(abs(labelledGensC - genVal) < minLabelSpacing)
        continue;
    end

    peakTheta = 2 * pi * genVal / period;
    if invertCircular
        peakR  = 1 - pks(li);
        labelR = max(0, peakR - 0.05);
    else
        peakR  = pks(li);
        labelR = min(1, peakR + 0.05);
    end
    polarplot(peakTheta, peakR, 'r.', 'MarkerSize', 3);

    text(peakTheta, labelR, sprintf('%.1f', genVal), ...
        'FontSize', 7, 'Color', [0.8 0 0], ...
        'HorizontalAlignment', 'center');

    labelledGensC = [labelledGensC, genVal]; %#ok<AGROW>
    labelledC = labelledC + 1;
    if labelledC >= nLabels
        break;
    end
end
hold off;

%% === Console output: top peaks ===
% Show only generators in [0, period/2] to avoid duplication (gen and
% period - gen produce the same scale and hence the same SPCS; any tiny
% difference is due to floating-point arithmetic). The complement
% (period - gen) is shown in parentheses.

% For each peak, take the best SPCS of gen and period-gen
halfPeriod = period / 2;
fprintf('\nTop peaks (generator values with highest SPCS):\n');
fprintf('%-14s  %-14s  %s\n', 'gen (cents)', '(complement)', 'SPCS');
fprintf('%s\n', repmat('-', 1, 42));
labelledGensT = [];
printed = 0;
for li = 1:numel(pks)
    genVal = genRange(locs(li));

    % Fold to [0, period/2]
    if genVal > halfPeriod
        genVal = period - genVal;
    end
    complement = period - genVal;

    % Skip if too close to an already-printed generator
    if ~isempty(labelledGensT) && any(abs(labelledGensT - genVal) < minLabelSpacing)
        continue;
    end

    % Best SPCS of gen and its complement (should be identical up to
    % floating-point error)
    idxGen  = round(genVal / genStep) + 1;
    idxComp = round(complement / genStep) + 1;
    idxGen  = max(1, min(idxGen, numel(s)));
    idxComp = max(1, min(idxComp, numel(s)));
    bestSpcs = max(s(idxGen), s(idxComp));

    fprintf('%-14.1f  (%-10.1f)  %.3f\n', genVal, complement, bestSpcs);
    labelledGensT = [labelledGensT, genVal]; %#ok<AGROW>
    printed = printed + 1;
    if printed >= nLabels
        break;
    end
end

File: python/demos/demo_gen_chain_spcs.py
"""demo_gen_chain_spcs.py

Spectral pitch class similarity (SPCS) of generator-chain tunings
against a reference chord, as the generator interval is swept
continuously.

A generator-chain of cardinality n is a scale formed by stacking n
copies of a generating interval, reduced modulo the period:
  pitches = mod(arange(n) * gen, period)

Produces both a linear and a circular (polar) plot.

Examples of this type of plot appear as Examples 6.4–6.5 / Figures 5–7
in:
  Milne, A. J., Sethares, W. A., Laney, R., & Sharp, D. B. (2011).
  Modelling the similarity of pitch collections with expectation tensors.
  Journal of Mathematics and Music, 5(1), 1-20.

Port of demo_genChainSpcs.m from the MATLAB Music Perception Toolbox v2.

Requires: matplotlib (pip install matplotlib)
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import find_peaks

import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))
import mpt

# ===================================================================
#  User-adjustable parameters
# ===================================================================

# Reference chord (in cents)
ref_pitches = [0, 1200 * np.log2(5 / 4), 1200 * np.log2(6 / 4)]
ref_name = '4:5:6 JI major triad'

# Generator-chain parameters
n_tones = 19
gen_step = 0.5   # step size for generator sweep (cents)

# Expectation tensor parameters
sigma = 10
r = 2
is_rel = True
is_per = True
period = 1200

# Display options
invert_circular = True   # True: higher similarity towards centre
n_labels = 15
min_label_spacing = 9    # minimum spacing between labels (cents)

# ===================================================================
#  Build pitch matrices (0 to period/2 only)
# ===================================================================

half_range = np.arange(0, period / 2 + gen_step / 2, gen_step)
n_half = len(half_range)

# Reference chord: same for every row
pitches_a = np.tile(ref_pitches, (n_half, 1))

# Generator-chain pitch sets
pitches_b = np.full((n_half, n_tones), np.nan)
for i, gen in enumerate(half_range):
    pitches_b[i, :] = np.mod(np.arange(n_tones) * gen, period)

# ===================================================================
#  Compute similarities
# ===================================================================

print(f"Computing SPCS of {n_tones}-tone generator-chains "
      f"(gen = 0 to {period / 2:.1f}, step {gen_step}) "
      f"against {ref_name}...")

s_half = mpt.batch_cos_sim_exp_tens(
    pitches_a, pitches_b, sigma, r, is_rel, is_per, period,
    verbose=False,
)
s_half = np.round(s_half, 3)
print("Done.")

# Reflect to full range [0, period)
gen_range = np.arange(0, period, gen_step)
s = np.concatenate([s_half, s_half[-2:0:-1]])

# Trim or pad to match gen_range
if len(s) < len(gen_range):
    s = np.concatenate([s, np.full(len(gen_range) - len(s), s[-1])])
elif len(s) > len(gen_range):
    s = s[:len(gen_range)]

# ===================================================================
#  Find peaks
# ===================================================================

pk_idx, pk_props = find_peaks(s, prominence=0.01)
pk_vals = s[pk_idx]

# Sort by height (descending)
sort_order = np.argsort(pk_vals)[::-1]
pk_idx = pk_idx[sort_order]
pk_vals = pk_vals[sort_order]

# ===================================================================
#  Linear plot
# ===================================================================

fig1, ax1 = plt.subplots(figsize=(12, 5))

ax1.plot(gen_range, s, color=(0.2, 0.2, 0.6), linewidth=0.8)
ax1.set_xlabel('Generator (cents)')
ax1.set_ylabel('Spectral pitch class similarity')
ax1.set_title(f'SPCS of {n_tones}-tone generator-chains with {ref_name}\n'
              f'(r = {r}, is_rel = {is_rel}, σ = {sigma} cents, '
              f'period = {period} cents)')
ax1.set_xlim(0, period)
ax1.grid(True, alpha=0.3)

labelled_gens = []
labelled = 0
for li in range(len(pk_idx)):
    gen_val = gen_range[pk_idx[li]]
    if any(abs(gen_val - g) < min_label_spacing for g in labelled_gens):
        continue
    ax1.plot(gen_val, pk_vals[li], 'r.', markersize=3)
    ax1.annotate(f'  {gen_val:.1f}', (gen_val, pk_vals[li]),
                 fontsize=7, color=(0.8, 0, 0), ha='left', va='bottom')
    labelled_gens.append(gen_val)
    labelled += 1
    if labelled >= n_labels:
        break

plt.tight_layout()

# ===================================================================
#  Circular plot
# ===================================================================

theta = 2 * np.pi * gen_range / period
theta_c = np.append(theta, theta[0])
s_c = np.append(s, s[0])

if invert_circular:
    r_c = 1 - s_c
else:
    r_c = s_c

fig2, ax2 = plt.subplots(figsize=(8, 8), subplot_kw={'projection': 'polar'})

ax2.plot(theta_c, r_c, color=(0.2, 0.2, 0.6), linewidth=1)
ax2.set_theta_zero_location('N')
ax2.set_theta_direction(-1)  # clockwise
ax2.set_rlim(0, 1)

# Tick labels in cents
n_ticks = 12
tick_angles = np.linspace(0, 2 * np.pi * (1 - 1 / n_ticks), n_ticks)
tick_labels = [f'{a * period / (2 * np.pi):.0f}' for a in tick_angles]
ax2.set_xticks(tick_angles)
ax2.set_xticklabels(tick_labels)
ax2.set_yticklabels([])

title_extra = 'centre = most similar' if invert_circular else 'perimeter = most similar'
ax2.set_title(f'SPCS of {n_tones}-tone generator-chains with {ref_name}\n'
              f'(r = {r}, is_rel = {is_rel}, σ = {sigma}, '
              f'period = {period}; {title_extra})',
              pad=20)

# Label top peaks (ranked by height, minimum spacing in cents)
labelled_gens_c = []
labelled_c = 0
for li in range(len(pk_idx)):
    gen_val = gen_range[pk_idx[li]]
    if any(abs(gen_val - g) < min_label_spacing for g in labelled_gens_c):
        continue
    peak_theta = 2 * np.pi * gen_val / period
    if invert_circular:
        peak_r = 1 - pk_vals[li]
        label_r = max(0, peak_r - 0.05)
    else:
        peak_r = pk_vals[li]
        label_r = min(1, peak_r + 0.05)
    ax2.plot(peak_theta, peak_r, 'r.', markersize=3)
    ax2.annotate(f'{gen_val:.1f}', (peak_theta, label_r),
                 fontsize=7, color=(0.8, 0, 0), ha='center', va='center')
    labelled_gens_c.append(gen_val)
    labelled_c += 1
    if labelled_c >= n_labels:
        break

plt.tight_layout()

# ===================================================================
#  Console output: top peaks
# ===================================================================

half_period = period / 2
print(f"\nTop peaks (generator values with highest SPCS):")
print(f"{'gen (cents)':<14s}  {'(complement)':<14s}  {'SPCS'}")
print('-' * 42)
labelled_t = []
printed = 0
for li in range(len(pk_idx)):
    gen_val = gen_range[pk_idx[li]]
    if gen_val > half_period:
        gen_val = period - gen_val
    complement = period - gen_val

    if any(abs(gen_val - g) < min_label_spacing for g in labelled_t):
        continue

    idx_gen = max(0, min(int(round(gen_val / gen_step)), len(s) - 1))
    idx_comp = max(0, min(int(round(complement / gen_step)), len(s) - 1))
    best_spcs = max(s[idx_gen], s[idx_comp])

    print(f"{gen_val:<14.1f}  ({complement:<10.1f})  {best_spcs:.3f}")
    labelled_t.append(gen_val)
    printed += 1
    if printed >= n_labels:
        break

plt.show()
